doc/design/dynamic_tls.md

```diff
@@ -0,0 +1,66 @@
+# Dynamic TLS Certs for etcd Cluster in etcd Operator
+
+Author: Eric Chiang(eric.chiang@coreos.com), Hongchao Deng(hongchao.deng@coreos.com)
+
+Motivation:
+- Simplify TLS cert generation for etcd cluster.
+- Self-hosted etcd wants to depend on Node IP, not hostnames.
+
+Requirements:
+- Dynamically generate certs per etcd pod
+- Able to use host IP in self hosted.
+
+### Option 1: K8S CSR, generate on etcd pod init container
+
+[Kubernetes certificates API][k8s_csr] provides the ability to send and approve CSR.
+
+Workflow:
+- etcd operator creates an etcd pod.
+  There is a separate service account for etcd pod to access  CSR API.
+- etcd pod first runs an init container to send a CSR and then wait for approval.
+- An external process should verify the CSR and approve it if correct.
+- Once approved, etcd pod gets the certs and bootstraps.
+
+Pros:
+- Hooks into existing k8s.
+- Delegates CA signing cert management. Not touching the signing certs and keys directly.
+
+Cons:
+- Share the same CA signing cert + key as k8s.
+- Requires etcd to be trusted by k8s cluster CA (probably only appropriate for self-hosted etcd).
+- Requires additional admin setup: approver, etcd pod service account.
+
+
+### Option 2: CA signing cert + key, generate on init etcd pod init container
+
+Add a secret reference to CA signing cert + key in EtcdCluster spec. Use it to generate certs for etcd pod.
+
+Workflow:
+- etcd operator creates an etcd pod with a volume mounted with CA secret.
+- etcd pod first runs an init container to use the CA cert + key to generate certs in an emptyDir.
+- etcd container then uses the generated certs in bootstrap.
+
+Pros:
+- Each etcd cluster has its own CA.
+
+Cons:
+- Inappropriate to give etcd pod access to the CA key?
+
+### Option 3: CA signing cert + key, etcd operator generates certs and secrets
+
+Similar to option 2, but instead of using init container, etcd operator generates certs and puts them into a secret. Then etcd operator creates an etcd pod with the secret mounted.
+
+The option is not feasible. The pod IP cannot be known before it is created. In self-hosted mode, even if we can know the node addresses, it is still a problem that the node could be deleted and new ones be added. Thus, we can’t specify the SAN until the pod is created.
```

Comment:
In self hosted mode the Pod IP is the node IP, right? So that is known ahead of time if the etcd-operator schedules the pod itself (like the daemonset controller)

Reply:
In theory, yeah. I think @hongchaodeng was worried about edge cases where the node IP suddenly changes before the pod is scheduled. I don't think this is much of a problem, though. 

If Option 2 is not an option due to security concerns, I'm more of a fan of Option 3 than 1. At least for now.

Reply:
> In theory, yeah. I think @hongchaodeng was worried about edge cases where the node IP suddenly changes before the pod is scheduled. I don't think this is much of a problem, though.

That sounds like a super-extreme edge case I don't see happening. At least, I think you'll have way worse problems if you're master node IP changes that frequently. I vote for calculating it based on node IP

Reply:
> I think @hongchaodeng was worried about edge cases where the node IP suddenly changes before the pod is scheduled.

Does that actually happen?

Reply:
That's not edge case. Let me clarify the concerns:

### Concern 1: Changing existing nodes

Etcd cluster running on 3 master nodes N1, N2, N3. We are going to replace N3 with N4. In etcd 3.2+, peer's cert SAN needs to include all peer addresses/IPs. Simply adding member of N4 won't be able to join N1 and N2.

### Concern 2: How to know the master nodes?

etcd operator do not know on which node the etcd pod is gonna end up until scheduler makes the decision, and thus do not know what SAN should be included.

If we want to know that, we need to include some scheduling logic in etcd operator.

Reply:
> peer's cert SAN needs to include all peer addresses/IPs. Simply adding member of N4 won't be able to join N1 and N2.

Can you expand on that? How does this affect Option 3 but not Option 1? Surely you're doing the same thing, i.e. generating a cert bound to a node IP.

> etcd operator do not know on which node the etcd pod is gonna end up until scheduler makes the decision, and thus do not know what SAN should be included.

I think Lucas's point is that we can infer it. The operator already has [a list of master nodes](https://github.com/coreos/etcd-operator/blob/c7131f581f00bf0901d24d23fc1768f7103ba01d/pkg/cluster/self_hosted.go#L33), so finding the ones that do not have etcd pods on should be simple. 

If the node is replaced in the seconds _after_ the certs are generated but _before_ the pod is scheduled, this of course won't work. But I think this is very rare.

Reply:
> Can you expand on that? How does this affect Option 3 but not Option 1?

ref: https://github.com/coreos/etcd/issues/8797

You are right. Option 1 will definitely not work at least for current 3.2+.

Reply:
Actually I made a mistake in understanding... Sorry for the confusion.

It's the certificate presented by remote peer: https://github.com/golang/go/blob/go1.9/src/crypto/tls/common.go#L169:2

So that means as long as the certificate for each peer includes itself, it should be fine.

Ignore the first concern.

Reply:
> If the node is replaced in the seconds after the certs are generated but before the pod is scheduled, this of course won't work. But I think this is very rare.

Yeah. We probably do not need to care about that at the beginning.
But this is automated system. And we should definitely need to handle that if we go with option 3.